src/sfizz/FilePool.h

```diff
@@ -346,40 +417,73 @@ class FilePool {
 
     // Signals
     volatile bool dispatchFlag { true };
-    volatile bool garbageFlag { true };
-    RTSemaphore dispatchBarrier;
-    RTSemaphore semGarbageBarrier;
+    RTSemaphore dispatchBarrier { 0 };
 
     // Structures for the background loaders
     struct QueuedFileData
     {
         QueuedFileData() noexcept {}
-        QueuedFileData(std::weak_ptr<FileId> id, FileData* data) noexcept
+        QueuedFileData(std::weak_ptr<FileId> id, const std::shared_ptr<FileData>& data) noexcept
         : id(id), data(data) {}
         std::weak_ptr<FileId> id;
-        FileData* data { nullptr };
+        std::shared_ptr<FileData> data { nullptr };
     };
 
     using FileQueue = atomic_queue::AtomicQueue2<QueuedFileData, config::maxVoices>;
     aligned_unique_ptr<FileQueue> filesToLoad;
 
     void dispatchingJob() noexcept;
-    void garbageJob() noexcept;
     void loadingJob(const QueuedFileData& data) noexcept;
     std::mutex loadingJobsMutex;
     std::vector<std::future<void>> loadingJobs;
     std::thread dispatchThread { &FilePool::dispatchingJob, this };
-    std::thread garbageThread { &FilePool::garbageJob, this };
 
-    SpinMutex garbageAndLastUsedMutex;
-    std::vector<FileId> lastUsedFiles;
-    std::vector<FileAudioBuffer> garbageToCollect;
+public:
+    struct GlobalObject
+    {
+        friend class sfz::FilePool;
+        friend struct sfz::FileData;
+        GlobalObject(size_t num_threads);
+
+        public:
+        ~GlobalObject();
+
+        public:
+        size_t getNumLoadedSamples() {
+            return loadedFiles.size() + preloadedFiles.size();
+        }
+
+        private:
+        volatile bool garbageFlag { true };
+        std::atomic<uint32_t> runningRender = { 0 };
+        std::chrono::time_point<std::chrono::high_resolution_clock> lastGarbageCollection_ = {};
+        std::unique_ptr<std::thread> garbageThread {};
+        // this semaphore should be static in mac.
+        static RTSemaphore semGarbageBarrier;
+        void garbageJob();
+
+        std::unique_ptr<ThreadPool> threadPool;
+
+        std::mutex loadedFilesMutex;
+        std::mutex preloadedFilesMutex;
+
+        // Preloaded data
+        absl::flat_hash_map<FileId, std::shared_ptr<FileData>> preloadedFiles;
```

Comment:
Hey, just a quick question: what made you change these to shared pointers? My concern is that this shared pointer is copied in `getFilePromise`, and this copy will most probably take a locking mutex, which I want to avoid on the realtime thread.